kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java

```diff
@@ -0,0 +1,290 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2021 Eurotech and/or its affiliates and others
+ * 
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Contributors:
+ *  Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.linux.position;
+
+import static java.lang.Math.toRadians;
+
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.eclipse.kura.linux.position.GpsDevice.Listener;
+import org.eclipse.kura.position.NmeaPosition;
+import org.osgi.util.measurement.Measurement;
+import org.osgi.util.measurement.Unit;
+import org.osgi.util.position.Position;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.taimos.gpsd4java.api.IObjectListener;
+import de.taimos.gpsd4java.backend.GPSdEndpoint;
+import de.taimos.gpsd4java.types.ATTObject;
+import de.taimos.gpsd4java.types.DeviceObject;
+import de.taimos.gpsd4java.types.DevicesObject;
+import de.taimos.gpsd4java.types.ENMEAMode;
+import de.taimos.gpsd4java.types.SKYObject;
+import de.taimos.gpsd4java.types.TPVObject;
+import de.taimos.gpsd4java.types.subframes.SUBFRAMEObject;
+
+public class GpsdPositionProvider implements PositionProvider, IObjectListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(GpsdPositionProvider.class);
+
+    private GPSdEndpoint gpsEndpoint;
+
+    private AtomicReference<Double> latitude = new AtomicReference<>();
+    private AtomicReference<Double> latitudeError = new AtomicReference<>();
+    private AtomicReference<Double> longitude = new AtomicReference<>();
+    private AtomicReference<Double> longitudeError = new AtomicReference<>();
+    private AtomicReference<Double> altitude = new AtomicReference<>();
+    private AtomicReference<Double> altitudeError = new AtomicReference<>();
+    private AtomicReference<Double> speed = new AtomicReference<>();
+    private AtomicReference<Double> speedError = new AtomicReference<>();
+    private AtomicReference<Double> course = new AtomicReference<>();
+    private AtomicReference<Double> courseError = new AtomicReference<>();
+    private AtomicReference<Double> timestamp = new AtomicReference<>();
+    private AtomicReference<ENMEAMode> mode = new AtomicReference<>();
```

Comment:
Could it make sense to define an internal "State" class containing all of these fields? In this way we can keep a single `AtomicReference<State>` here and have the whole state updated atomically. Otherwise it is possible that different fields come from different `TPVObject` instances due to race conditions.

Reply:
Looking at the current Gpsd4Java implementation it seems hard the race condition will happen, however it is a good idea. In the future we have to notice if a field is related to the others before add it in the internal state.